pkg/storage/unified/search/document.go

```diff
@@ -0,0 +1,170 @@
+package search
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"time"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/grafana/grafana/pkg/apimachinery/utils"
+	"github.com/grafana/grafana/pkg/services/store/kind/dashboard"
+	"github.com/grafana/grafana/pkg/storage/unified/resource"
+)
+
+type DocumentBuilderProvider interface {
+	GetDocumentBuilders(ctx context.Context) ([]resource.DocumentBuilderInfo, error)
+}
+
+// Replaced in enterprise with a version that includes stats
+func ProvideBuilders() DocumentBuilderProvider {
+	return &standardDocumentProvider{}
+}
+
+type standardDocumentProvider struct {
+	_ int // sealed
+}
+
+type defaultDocumentBuilder struct {
+	_ int // sealed
+}
+
+var (
+	_ resource.DocumentBuilder = &defaultDocumentBuilder{}
+	_ DocumentBuilderProvider  = &standardDocumentProvider{}
+)
+
+func (p *standardDocumentProvider) GetDocumentBuilders(ctx context.Context) ([]resource.DocumentBuilderInfo, error) {
+	return []resource.DocumentBuilderInfo{
+		{
+			Builder: &defaultDocumentBuilder{},
+		},
+		{
+			GroupResource: schema.GroupResource{
+				Group:    "dashboard.grafana.app",
+				Resource: "dashboards",
+			},
+
+			// This is a dummy example, and will need resolver setup for enterprise stats and and (eventually) data sources
+			Namespaced: func(ctx context.Context, namespace string, blob resource.BlobSupport) (resource.DocumentBuilder, error) {
+				lookup := dashboard.CreateDatasourceLookup([]*dashboard.DatasourceQueryResult{
+					// TODO, query data sources
+				})
+				return &DashboardDocumentBuilder{
+					Namespace: namespace,
+					Lookup:    lookup,
```

Comment:
since the dashboard list shows "Location" seems we will need a folderLookup here? so that we can get the folder name from the folder index (or a cache) by id? which means the folder index has to be loaded before dashboards.

Reply:
I'm not sure -- for sure as a first pass it is not required, and this gives us a place to add that if necessary.  

In searchV2, we added just the folder(uid) to the index -- and then sent full folder information for everything in the response as a side package.  see:
Backend:
https://github.com/grafana/grafana/blob/v11.3.0//pkg/services/searchV2/bluge.go#L704
Frontend:
https://github.com/grafana/grafana/blob/v11.3.0/public/app/features/search/service/types.ts

Given that *every* resource that needs to support folders will have this issue; perhaps we can look at supporting this in the frontend with an additional request.  Similar to how it will be solved to show user display information?

Reply:
Do we ever need to *sort* dashboards by the name of the folder it lives in?  If yes, then we need it in the index -- if not we can "join" it later.

Reply:
the question is how do we join it later.  we don't really need "sprinkles" here since we should have the folder names in the index ( if folders are indexed first )